src/tdigest.rs

```diff
@@ -0,0 +1,269 @@
+
+        // the following chunk of code does the same thing as search_centroids but it's based on quantile instead of mean and also it's much faster for the use case (most queries are going to be at the extremes, so linear search works out to be pretty fast)
+        let mut pos: usize;
+        let mut t: f32;
+        if q > 0.5 {
+            if q >= 1.0 {
+                return self_max;
+            }
+
+            pos = 0;
+            t = count_;
+
+            for (k, centroid) in self.centroids.iter().enumerate().rev() {
+                t -= centroid.weight as f32;
+
+                if rank >= t {
+                    pos = k;
+                    break;
+                }
+            }
+        } else {
+            if q <= 0.0 {
+                return self_min;
+            }
+
+            pos = self.centroids.len() - 1;
+            t = 0.0;
+
+            for (k, centroid) in self.centroids.iter().enumerate() {
+                if rank < t + centroid.weight as f32 {
+                    pos = k;
+                    break;
+                }
+
+                t += centroid.weight as f32;
+            }
+        }
+
+        // TODO: document this lmao
+        let mut delta = 0.0;
+        let mut min: f32 = self_min.clone();
+        let mut max: f32 = self_max.clone();
+
+        if self.centroids.len() > 1 {
+            if pos == 0 {
+                delta = self.centroids[pos + 1].mean - self.centroids[pos].mean;
+                max = self.centroids[pos + 1].mean;
+            } else if pos == (self.centroids.len() - 1) {
+                delta = self.centroids[pos].mean - self.centroids[pos - 1].mean;
+                min = self.centroids[pos - 1].mean;
+            } else {
+                delta = (self.centroids[pos + 1].mean - self.centroids[pos - 1].mean) / 2.0;
+                min = self.centroids[pos - 1].mean;
+                max = self.centroids[pos + 1].mean;
+            }
+        }
+
+        let value = self.centroids[pos].mean + ((rank - t) / self.centroids[pos].weight as f32 - 0.5) * delta;
+        Self::clamp(value, min, max)
+    }
+    
+    // wanted to add a sorted version of this but i couldn't think of a version to save that much time and this is pretty fast already. if you wanted to make it super duper fast that's what you would do i guess
+    /// Merges an unsorted vector of values into the T-Digest.
+    pub fn merge_vec(&mut self, v: Vec<f32>) {
+        self.centroids.extend(Self::vals_to_centroids(v));
+        self.centroids.sort_by(|a, b| a.mean.partial_cmp(&b.mean).unwrap());
+        self.cluster_centroids();
+    }
+
+    /// Adds a single value into the T-Digest, using amortization.
+    /// Does batch updates to the T-Digest every 32 added values. In the meantime, the structure will not change.
+    pub fn add_value(&mut self, value: f32) {
+        self.buffer[self.index as usize] = value;
+        self.index += 1;
+        if self.index == 32 {
```

Comment:
this is intentional, it's supposed to be used with the knowledge that the structure won't be completely up-to-date at any given time but individual merges will be much faster. making it so that the buffer is drained before any reads would mostly defeat the purpose, since you're just relocating the latency that comes with merging values (it's the slowest operation by far).
i did plan on adding more descriptive docstrings to make this clear later on

Reply:
You could expose a method that drains the internal buffer manually though, so that if a user needs exact calculations, they can call the method before reading statistics